capsules/src/virtual_rng.rs

```diff
@@ -0,0 +1,167 @@
+// Virtualizer for the RNG
+use core::cell::Cell;
+use kernel::common::cells::OptionalCell;
+use kernel::common::{List, ListLink, ListNode};
+use kernel::hil::rng::{Client, Continue, Rng};
+use kernel::ReturnCode;
+
+#[derive(Copy, Clone, PartialEq)]
+enum Op {
+    Idle,
+    Get,
+}
+
+// Struct to manage multiple rng requests
+pub struct MuxRngMaster<'a, R: Rng<'a>> {
+    rng: &'a dyn Rng<'a>,
+    devices: List<'a, VirtualRngMasterDevice<'a, R>>,
+    inflight: OptionalCell<&'a VirtualRngMasterDevice<'a, R>>,
+}
+
+impl<'a, R: Rng<'a>> MuxRngMaster<'a, R> {
+    pub const fn new(rng: &'a dyn Rng<'a>) -> MuxRngMaster<'a, R> {
+        MuxRngMaster {
+            rng: rng,
+            devices: List::new(),
+            inflight: OptionalCell::empty(),
+        }
+    }
+
+    // TODO: return value is hacky way to surface return value from get
+    fn do_next_op(&self) -> ReturnCode {
+        if self.inflight.is_none() {
+            let mnode = self
+                .devices
+                .iter()
+                .find(|node| node.operation.get() != Op::Idle);
+            mnode.map(|node| {
+                let op = node.operation.get();
+
+                // Need to set idle here in case callback changes state
+                node.operation.set(Op::Idle);
+                self.inflight.set(node);
+
+                match op {
+                    Op::Get => {
+                        let client = node.client.take();
+                        match client {
+                            Some(p) => {
+                                // Set rng client to current node
+                                self.rng.set_client(p);
+                                node.client.insert(client);
+                                self.rng.get()
```

Comment:
we should only set the current node as `inflight` if `get()` returns `SUCCESS`. Otherwise no callback will be generated and the RNG will become unusable.